resources/git-rebase.md

```diff
@@ -0,0 +1,81 @@
+
+Using the interactive mode when performing a rebase helps as well. Interactive mode can be used by adding the -i flag to your rebase command like so:
+
+ `git rebase -i <nameOfBranchToRebaseIn>`
+
+Using interactive mode when rebasing lets you see the commits that you're rebasing, and allows you to edit and squash commits.
+
+## What is git squash?
+Git squash allows you to take several commits and combine them into a single commit. Say you've finished a feature - styling a nav bar - and your git history looks like this:
+```
+3f0e9cdcda init commit
+9845111670 in progress
+cd362d2f52 still working
+300571e57c styled nav bar
+```
+
+To keep your git history clean and easy to read, you can combine all of these commits into a single commit and update your commit message. You can do this by rebasing your current branch into itself, and then squashing your commits together, like this:
+
+`git rebase -i <nameOfBranchYouAreOn>~4`
+
+The reason "4" is added to the end of the rebase message is because your branch currently has 4 commits that you want to make into a single commit. After you run this command, git will show you the commits you've selected:
+
+```
+pick 3f0e9cdcda init commit
+pick 9845111670 in progress
+pick cd362d2f52 still working
+pick 300571e57c styled nav bar
+```
+
+You can squash your commits by keeping the first commit as is, and selecting 'squash' for the other commits like so:
+
+```
+pick 3f0e9cdcda init commit
+s 9845111670 in progress
+s cd362d2f52 still working
+s 300571e57c styled nav bar
+```
+
+Save your choices:
+`cmd + s`, then `cmd+ w`
+
+Once you've done this, git will give you the option of editing the commit messages for all of the commits you've changed. You can edit your commit messages into a single commit that describes the work you've done on the branch. When you're finished editing the messages, save and quit your editor.
+
+At this point you can push your new, squashed changes. You'll likely need to force push, so rather than:
+`git push`
+
+you'll need:
+
+`git push -f`
+
+You use the -f flag to force update your git history on your remote branch after you rebase. If you don't force push your code, you'll get merge conflicts, and no one likes dealing with those.
+
+## When shouldn't I rebase?
+Once you've pushed your code to your team's repo, don't rebase anymore. If your code is in the repo, other people on your team may have pulled your work and rebased their branches. Rebasing at this point can cause messiness and merge conflicts.
+
+Similairly, once you've put up a PR and someone on your team has reviewed it and requested changes, don't rebase after you make these changes. You want to keep the history of your commits intact at this point so your reviewers can easily follow the work you've updated.
```

Comment:
I wanted to add something like "_you may need to rebase after your PR is approved in order to allow the code to be merged_" here. But, this one can get a bit complicated in practice, depending on the team process.  

On my current team, for example, we have Jenkins running on PRs and it will fail if there are outdated changes from master, and the reviewers will withhold approval. I'm not sure how to capture things like this aside from perhaps noting that the exact flow will vary from team to team.

Reply:
Definitely. I'll note that there are other times to rebase as well (like after your PR is approved but your branch now has merge conflicts) but that the way this can be handled will depend on your team norms.